Add LRU implementation

nicolov opened the pull request:
Add a size-aware LRU cache that we can use to limit the on-disk cache size. This PR only adds the generic cache, but does not wire it in yet.

This PR includes my logging changes, do you want to review those first?

buchgr commented on cache/utils_test.go:
```diff
@@ -34,3 +35,10 @@ func createTmpCacheDirs(t *testing.T) string {
+func newSilentLogger() (l log.Logger) {
+	l = *log.New(ioutil.Discard, "", 0)
+	return
+}
```
new line?

buchgr commented on main.go:
```diff
@@ -29,8 +30,10 @@ func main() {
+	errorLogger := log.New(os.Stdout, "", log.Ldate | log.Ltime | log.LUTC)
 	e := cache.NewEnsureSpacer(0.95, 0.5)
-	h := cache.NewHTTPCache(*dir, *maxSize*1024*1024*1024, e)
+	h := cache.NewHTTPCache(*dir, *maxSize*1024*1024*1024, e, *accessLogger, *errorLogger)
```
pass by reference?

nicolov replied:
made it an interface

buchgr commented on cache/sizedlru/sizedlru.go:
```diff
@@ -0,0 +1,111 @@
+}
+
+// SizedLRU is an LRU cache that will keep its total size below maxSize by evicting
+// items. If maxSize is zero, eviction is disabled.
```
I am thinking we might not even want to support this. IMHO there's no use case for an unbounded cache and one can always set maxSize to a really big value (or disk size). Wdyt?

nicolov replied:
removed

buchgr commented on cache/sizedlru/sizedlru.go:
```diff
@@ -0,0 +1,111 @@
+
+// Add adds a (key, value) to the cache, evicting items as necessary. Add returns false (
+// and does not add the item) if the item size is larger than the maximum size of the cache.
+func (c *SizedLRU) Add(key Key, value SizedItem) (ok bool) {
```
I am thinking ahead when using this actually implement eviction. We would need to return a slice of SizedItem that needs to be deleted from the disk?

nicolov replied:
a callback?

buchgr commented on cache/http.go:
```diff
@@ -90,22 +105,36 @@ func ensureDirExists(path string) {
+		// Parse the client ip
+		clientIp, _, err := net.SplitHostPort(r.RemoteAddr)
+		if err != nil {
+			h.errorLogger.Fatal("Failed parsing RemoteAddr %s: %s", r.RemoteAddr, err.Error())
```
Use `h.errorLogger.Fatalf`

buchgr replied:
Actually, I am thinking using `.Fatal` here is wrong. A single request would crash the whole server (this is why go test fails on CI). Can we just log `r.RemoteAddr` as is and thus not even encounter this error case?

nicolov replied:
removed the Fatal call

buchgr commented:
@nicolov can you please run `bazel run //:gazelle` to automatically generated a `BUILD` file for the `sizedlru` package.

buchgr commented on cache/sizedlru/sizedlru.go:
```diff
@@ -0,0 +1,111 @@
+
+// SizedLRU is an LRU cache that will keep its total size below maxSize by evicting
+// items. If maxSize is zero, eviction is disabled.
+// SizedLRU is not thread-safe.
```
Are you planning on wrapping every call with a mutex? I could see this becoming a bottleneck. Wdyt?

nicolov replied:
Correct, but the critical section should only contain a few pointer swaps for the linked list and an hashtable lookup/update. This pales in comparison with all the work for setting up the HTTP connection and actually transfering the file. I'll be sure to test this with 3-4 gbit/s throughput.

buchgr replied:
Yeah it's probably premature optimization. I am thinking we need benchmarks :-).

nicolov replied:
Each request would:
1. lock the mutex
2. get the element from the lru, updating the eviction list
3. unlock the mutex
4. actually serve the file from disk

I would expect that the locking would be relatively unexpensive.

buchgr commented on main.go:
```diff
@@ -29,8 +30,10 @@ func main() {
 	}
 
+	accessLogger := log.New(os.Stdout, "", log.Ldate | log.Ltime | log.LUTC)
+	errorLogger := log.New(os.Stdout, "", log.Ldate | log.Ltime | log.LUTC)
```
`os.Stderr`?

nicolov replied:
I had one job..

buchgr commented on cache/sizedlru/sizedlru.go:
```diff
@@ -0,0 +1,111 @@
+// Package sizedlru implements an LRU cache that bounds the total space occupied
+// by the items it contains.
+
+package sizedlru
```
Maybe just call it lru?

nicolov replied:
I actually moved it into the main package.

nicolov commented on cache/http.go:
```diff
@@ -40,13 +43,25 @@ type statusPageData struct {
+// accessLogger will print one line for each HTTP request to the cache.
+// errorLogger will print unexpected server errors. Inexistent files and malformed URLs will not
+// be reported.
+func NewHTTPCache(cacheDir string, maxBytes int64, ensureSpacer EnsureSpacer, accessLogger log.Logger, errorLogger log.Logger) HTTPCache {
```
I tried to match what you did for the EnsureSpacer. I still haven't figured out a good rule of thumb.

nicolov replied:
I'll just create an interface for it.

nicolov commented:
I addressed the points in the review, with the exception of the eviction callback. I'll add it in a follow-up PR, since this is already a good amount of code to merge. Btw, I'm on ubuntu 18.04/Bazel 0.11 and can't get the tests to pass using bazel. I get an empty log, and nothing is streamed to stdout. go test works. Which versions are you on?

buchgr commented:
@nicolov I am on Google's Debian based Linux [1]. We are testing 14.04 and 16.04 but not 18.04 on our CI. I ll open an issue for Bazel to support 18.04.


[1] http://www.zdnet.com/article/google-moves-to-debian-for-in-house-linux-desktop/
